core/vm/contracts.go

```diff
@@ -896,6 +897,45 @@ func (c *blake2F) Run(input []byte, caller common.Address, evm *EVM, gas uint64)
 	return output, gas, nil
 }
 
+// ed25519Verify implements a native Ed25519 signature verification.
+type ed25519Verify struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+func (c *ed25519Verify) RequiredGas(input []byte) uint64 {
+	const sha2_512WordLength = 64
+
+	// round up to next whole word
+	lengthCeil := len(input) + sha2_512WordLength - 1
+	words := uint64(lengthCeil / sha2_512WordLength)
+	return params.Ed25519VerifyGas + params.Sha2_512BaseGas + (words * params.Sha2_512PerWordGas)
+}
+
+func (c *ed25519Verify) Run(input []byte, caller common.Address, evm *EVM, gas uint64) ([]byte, uint64, error) {
+	gas, err := debitRequiredGas(c, input, gas)
+	if err != nil {
+		return nil, gas, err
+	}
+
+	// Setup success/failure return values
+	var fail32byte, success32Byte = true32Byte, false32Byte
+
+	// Check if all required arguments are present
+	if len(input) < 96 {
+		return fail32byte, gas, nil
```

Comment:
shouldn't we return an error here?

Reply:
This is the same thing as in CIP-20, I think returning an error here aborts the TX. From the spec, it seems like invalid inputs are fine and you instead return an error inside the evm. It seems like there's a little more usage of returning and out of bound error, but given that the gas metering has to work for any input, I'm pretty fine with this return convention.

Reply:
returning an error here will revert the precompile execution and the contract that called it will receive a signal that it reverted. The CIP authors decided to return a failure code instead of reverting with an error. This way, future failure codes can be added to signal precise reasons the validation failed, without being married to any solidity-level error handling system. I would think that the `UsingPrecompiles` interface to this would (by default) revert on failure